Applications/Spire/Source/Ui/KeyFilterPanel.cpp

```diff
@@ -0,0 +1,103 @@
+#include "Spire/Ui/KeyFilterPanel.hpp"
+#include "Spire/Ui/AnyInputBox.hpp"
+#include "Spire/Ui/CustomQtVariants.hpp"
+#include "Spire/Ui/KeyInputBox.hpp"
+
+using namespace boost::signals2;
+using namespace Spire;
+
+struct KeyInputBoxValueModel : ValueModel<QKeySequence> {
+  std::shared_ptr<KeySequenceValueModel> m_source;
+  std::shared_ptr<AnyListModel> m_matches;
+  std::unordered_set<QString> m_matches_set;
+  std::vector<QString> m_matches_list;
+  scoped_connection m_matches_connection;
+
+  KeyInputBoxValueModel(std::shared_ptr<KeySequenceValueModel> source,
+      std::shared_ptr<AnyListModel> matches)
+      : m_source(std::move(source)),
+        m_matches(std::move(matches)),
+        m_matches_connection(m_matches->connect_operation_signal(
+          std::bind_front(&KeyInputBoxValueModel::on_operation, this))) {
+    for(auto i = 0; i < m_matches->get_size(); ++i) {
+      add_match(i);
+    }
+  }
+
+  QValidator::State get_state() const override {
+    return m_source->get_state();
+  }
+
+  const Type& get() const override {
+    return m_source->get();
+  }
+
+  QValidator::State test(const Type& value) const {
+    return m_source->test(value);
+  }
+
+  QValidator::State set(const Type& value) {
+    if(m_matches_set.contains(value.toString())) {
+      return QValidator::State::Invalid;;
+    }
+    return m_source->set(value);;
+  }
+
+  connection connect_update_signal(
+      const UpdateSignal::slot_type& slot) const override {
+    return m_source->connect_update_signal(slot);
```

Comment:
Because ```KeyInputBoxValueModel``` doesn't own ```m_source```, it can not just directly forward slots to ```m_source```. When ```KeyInputBoxValueModel``` is destroyed, all connections have to be destroyed as well.